### PR TITLE
feat(builders): add `AttachmentBuilder` and `UnsafeAttachmentBuilder`

### PR DESCRIPTION
**Pull request by Mateo-tem**
**Please describe the changes this PR makes and why it should be merged:**

I opened a PR recently (which was merged) to change `MessageAttachment` to `Attachment`.

- https://github.com/discordjs/discord.js/pull/7691

All was good, but I notice that some names were changed and some builders were added to the builders package.

**So, this PR adds `AttachmentBuilder` and `UnsafeAttachmentBuilder` to the builders package.**

(If something is wrong, please let me know)

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: packages/discord.js/src/structures/Attachment.js [hunk 1]
```diff
@@ -1,152 +1,103 @@
...
-      /**
-       * The media type of this attachment
-       * @type {?string}
-       */
-      this.contentType = data.content_type;
-    } else {
-      this.contentType ??= null;
-    }
+  /**
+   * The Proxy URL to this attachment
```
Comment (vitorlops) [inline packages/discord.js/src/structures/Attachment.js:80]:
```suggestion
   * The proxy URL to this attachment
```

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+			/**
+			 * The URL to this attachment
+			 * @type {string}
+			 */
+			this.url = data.url;
+		}
+
+		if ('proxy_url' in data) {
+			/**
+			 * The Proxy URL to this attachment
```
Comment (vitorlops) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:111]:
```suggestion
			 * The proxy URL to this attachment
```

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+import type { APIAttachment, Snowflake } from 'discord-api-types/v10';
+
+export type BufferResolvable = typeof ArrayBuffer | string;
+/**
+ * Represents an attachment.
+ */
+export class UnsafeAttachmentBuilder {
+	public readonly data: APIAttachment;
+	public attachment: BufferResolvable | Stream;
+	public contentType?: string | null;
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:11]: should be snake_cased instead

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	public description?: string | null;
+	public ephemeral?: boolean;
+	public height?: number | null;
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:14]: Builders cannot set this

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	public id!: Snowflake;
+	public name: string | null;
+	public proxyURL!: string;
+	public size!: number;
+	public url!: string;
+	public width?: number | null;
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:20]: Builders cannot set these

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
+import type { Stream } from 'node:stream';
+import type { APIAttachment, Snowflake } from 'discord-api-types/v10';
+
+export type BufferResolvable = typeof ArrayBuffer | string;
+/**
+ * Represents an attachment.
+ */
+export class UnsafeAttachmentBuilder {
+	public readonly data: APIAttachment;
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:9]: This won't be needed

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+			while (this.spoiler && typeof this.name === 'string') {
+				this.name = this.name.slice('SPOILER_'.length);
+			}
+			return this;
+		}
+		this.name ? (this.name = `SPOILER_${this.name}`) : (this.name = null);
+		return this;
+	}
+
+	private _patch(data: APIAttachment) {
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:86]: This method is not needed in builders

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+	/**
+	 * Sets the file of this attachment.
+	 * @param {BufferResolvable|Stream} attachment The file
+	 * @param {string} [name] The name of the file, if any
+	 * @returns {Attachment} This attachment
+	 */
+	public setFile(attachment: BufferResolvable | Stream, name: string | null) {
+		this.attachment = attachment;
+		this.name ? (this.name = name) : (this.name = null);
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:54]: Aren't file names required? 👀

File: packages/builders/src/messages/attachment/UnsafeAttachment.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
+import type { Stream } from 'node:stream';
+import type { APIAttachment, Snowflake } from 'discord-api-types/v10';
+
+export type BufferResolvable = typeof ArrayBuffer | string;
```
Comment (vladfrangu) [inline packages/builders/src/messages/attachment/UnsafeAttachment.ts:4]: 1. Small code style

```suggestion
export type BufferResolvable = typeof ArrayBuffer | string;

```

---

2. Are you sure its `typeof ArrayBuffer` and not `ArrayBuffer`?

File: packages/discord.js/src/structures/Attachment.js [hunk 10]
```diff
@@ -1,152 +1,103 @@
...
-   * @param {string} description The description of the file
-   * @returns {Attachment} This attachment
+   * The attachment's id
+   * @type {Snowflake}
    */
-  setDescription(description) {
-    this.description = description;
-    return this;
+  get id() {
+    return this.data.id ?? null;
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Attachment.js:28]:
```suggestion
    return this.data.id;
```

File: packages/discord.js/src/structures/Attachment.js [hunk 11]
```diff
@@ -1,152 +1,103 @@
...
-   * @returns {Attachment} This attachment
+   * The media type of this attachment
+   * @type {?string}
    */
-  setFile(attachment, name = null) {
-    this.attachment = attachment;
-    this.name = name;
-    return this;
+  get contentType() {
+    return this.data.content_type ?? this.data.contentType ?? null;
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Attachment.js:36]: data.contentType will never exist

File: packages/discord.js/src/structures/Attachment.js [hunk 12]
```diff
@@ -1,152 +1,103 @@
...
-      this.contentType = data.content_type;
-    } else {
-      this.contentType ??= null;
-    }
+  /**
+   * The proxy URL to this attachment
+   * @type {string}
+   */
+  get proxyURL() {
+    return this.data.proxy_url ?? this.data.proxyURL ?? null;
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Attachment.js:84]: Ditto, camelCase will never exist from API data

File: packages/discord.js/src/structures/Attachment.js [hunk 13]
```diff
@@ -158,8 +109,36 @@ class Attachment {
     return Util.basename(this.url ?? this.name).startsWith('SPOILER_');
   }
 
+  /**
+   * Creates a new attachment builder from JSON data
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Attachment.js:113]: This doesn't return a builder... This returns another Attachment class

**Review by vladfrangu — DISMISSED** [review]
Is this even something we want to add to builders? This is the only structure that doesn't fully make sense, since the api data doesn't match what you send (+ the extra features d.js adds on top of the class). I think this is the first structure that's making me more inclined to leaving it in d.js only, but still splitting it up into a builder and structure class (like embeds)

**Comment by Mateo-tem** [comment]
Yeah, the Attachment structure is confusing, but I will try to find the right one (to send).

**Pull request closed without merging**